File: C++/0007-reverse-integer.cpp
class Solution {
public:
    int reverse(int x) {
        int ret = 0;
        
        while (x != 0) {
            int digit = x % 10;
            
            // Check for overflow before multiplying or adding
            if (ret > INT_MAX / 10 || (ret == INT_MAX / 10 && digit > 7))
                return 0;
            if (ret < INT_MIN / 10 || (ret == INT_MIN / 10 && digit < -8))
                return 0;
            
            ret = ret * 10 + digit;
            x /= 10;
        }
        
        return ret;
    }
};

File: JavaScript/0007-reverse-integer.js
var reverse = function(x) {
    let result = 0;

    while (x !== 0) {
        let digit = x % 10;
        // Truncate x toward zero (to handle negatives)
        x = (x / 10) | 0;

        // Check for 32-bit overflow
        if (result > 214748364 || (result === 214748364 && digit > 7)) return 0;
        if (result < -214748364 || (result === -214748364 && digit < -8)) return 0;

        result = result * 10 + digit;
    }

    return result;
};
